_posts/2020-05-18-microservice-architectures.md

```diff
@@ -0,0 +1,80 @@
+
+But how do we characterise a microservice?  We do not characterise based on the number of lines of code, but rather on [bounded context](https://martinfowler.com/bliki/BoundedContext.html).  
+
+For me, microservices:
+* Are event driven
+* Are domain oriented
+* Decentralise data management
+
+## Event Driven
+
+Serverless computing is event driven at it's core.  Whether invoking [functions](https://docs.aws.amazon.com/lambda/latest/dg/lambda-invocation.html), [consuming services](https://docs.aws.amazon.com/rekognition/latest/dg/stored-video-lambda.html) or [building integrations](https://aws.amazon.com/eventbridge/).  
+
+Events comprise of two core elements:
+* Domain event - the important detail of the event and includes important attributes such as timestamp, and eventType.  We use the [cloud event specification](https://cloudevents.io/) for defining domain events.
+* Envelope - wraps the domain event and captures details of the event transport, such as [SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html) or [EventBridge](https://docs.aws.amazon.com/eventbridge/latest/userguide/aws-events.html) events.
+
+![clean architecture diagram showing onion of layers including handler, integration with ports and adapters, service and domain layers ](/img/microservice-architectures/clean-architecture.png)
+
+By following [clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) principles, separating envelope unmarshalling from acting on domain events, provides additional flexibility on your deployment architecture.  For example, you can use the same domain processing logic to handle synchronous requests sent via API gateway or asynchronously via SNS.
+
+Irrespective of whether you use [CQRS](https://martinfowler.com/bliki/CQRS.html), events come in two categories:
+* commands 
+* events
+
+Commands are a request to do something: `requestCheckout`, `updateBalance`, `createMyWidget`.  Commands are inherently transient, may be resubmitted multiple times and may/may not result in action being taken.  Events are a notification that something has happened, or a particular state: `checkoutCompleted`, `balanceUpdated`, `MyWidgetCreated`.  Events occurred at a point in in the past.  Depending on how you choose to interpret events, a key part of the semantics, event re-delivery may be possible.  For example, `accountCreated` may be interpreted as "an account has _just_ been created", or "an account was created at a certain point in time".  The former has implications on event re-delivery, whereas the latter is immutable and less affected by the temporal difference between when the actual creation occurred and when the event was generated.
```

Comment:
_Commands are inherently transient, may be resubmitted multiple times, and may/may not result in action being taken._

I feel like you need an oxford comma here.